### PR TITLE
Create smaller image, faster build times, rework dependencies

### PR DESCRIPTION
**Pull request by oivindoh**
Based on work in https://github.com/ostefano/docker-misp/pull/25

- Installs dependencies from MISP requirements.txt, with some additions added after checking out MISP. Fetch everything python (+ MISP) in the python builder stage
- Move away from github clones for dependencies where feasible
- Disable update from git functionality as this is an unwelcome workflow for containers (https://github.com/MISP/MISP/pull/9275 exists for removing an unnecessary error message this causes in diagnostics)
- Reduce image size drastically (sizes compressed)
  - linux/amd64: 450 MB
  - linux/arm64: 388 MB
- Build faster (because we mostly use wheels rather than source)
- Shuffle files around for less layering

### CODE REVIEW

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: Is it still TODO?

**Comment by ostefano** [comment]
Looks good overall. I would like to get clarity on pymisp though.
Correct solution right now would be to get the VERSION from the submodule and pip install that specific version.
If the PR on MISP (which doesnt exit yet) goes ahead, then we can just use requirements.txt.

**Comment by oivindoh** [comment]
> 

This feels a bit nasty, but could be added after submodule update?

```sh
            pymispversion=$(head -n1 PyMISP/pymisp/__init__.py | awk '{print $3}' | tr -d \')

            if [ -n "$pymispversion" ]; then
                sed -i "s/pymisp$/pymisp==$pymispversion/" requirements.txt
                cat requirements.txt
            fi
```

**Comment by ostefano** [comment]
@oivindoh thank you. I think it's ready to go. I want to test things locally first, and then I'll merge (give me a couple of days).

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: @oivindoh because of this line (which we might want to keep) the diagnostic page does not display the list of submodules. We should with other MISP devs what is the impact of this

**Comment by ostefano** [comment]
@oivindoh see comment above. Merged PR on `misp-module`, but will hold on here. 

Note that the diagnostic page is also complaining about the lack of python-magic.

File: server/Dockerfile
Comment (oivindoh) [reply server/Dockerfile]: I don't really think that's valuable information since we're disabling the update process for our static docker image anyway. Actually, we can also remove the git application itself since MISP is grabbed from prior stages now.

**Comment by oivindoh** [comment]
> 

RE: pythonmagic, added a commit to put that back.

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: Any reason for the $?

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: If the requirements file has lief>=0.14.1, lief>=0.13.1 will be added.
Can we define a mapping here with our pinnings, overriding those specified in requirements.txt ?

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: Seems to me there are 4 extra spaces

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: Seems to me that this block has 4 extra leading spaces

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: To be removed potentially

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: We should add `pip` here

Comment (oivindoh) [reply server/Dockerfile]: Trying very hard to come up with what my reasoning was, but coming up short here.

Comment (oivindoh) [reply server/Dockerfile:90]: Perhaps skip installing the debian package instead?

File: server/Dockerfile
Comment (ostefano) [reply server/Dockerfile]: Also, cybox is an explicit dependency that I see missing here

File: server/Dockerfile
Comment (oivindoh) [reply server/Dockerfile]: added

File: server/Dockerfile
Comment (oivindoh) [reply server/Dockerfile]: removed

File: server/Dockerfile
Comment (oivindoh) [reply server/Dockerfile]: fixed indentation

File: server/Dockerfile
Comment (oivindoh) [reply server/Dockerfile]: fixed indentation

File: server/Dockerfile
Comment (ostefano) [reply server/Dockerfile]: Let's remove it then.

**Comment by ostefano** [comment]
Thank you. The only issue left that I can see is the masking of dependency versions.
A mapping with our pinnings (to be added as override) could fix this.

**Comment by oivindoh** [comment]
Something like [this](https://github.com/ostefano/docker-misp/pull/27/commits/751153f28b310b19444e98f3a716021ec2758b09)?

**Comment by oivindoh** [comment]
Further reduced the image size (and hopefully improved build times; not sure how this works in github actions) by employing cache mounts for apt and bind-mounting /wheels rather than copying it into the final image.

**Comment by ostefano** [comment]
@oivindoh can we move this last optimization (cache rather than copy between layers) to another?

I would rather converge before discussing that.

**Comment by oivindoh** [comment]
> 

Reverted that change  as a standard commit

File: docker-bake.hcl
Comment (ostefano) [inline docker-bake.hcl]: I think there is a bit of confusion.

`null` is for variables that need to be set.
`Empty string` is for variables that can be initialized as empty.

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: I am not sure this works as expected.

Imagine the case where `requirements.txt` has a dependency `dep=0.1`, and the Dockerfile has `dep=0.5`.
Both would be added. I guess the grep should work only on the name of the dependency.

File: template.env
Comment (ostefano) [inline template.env:10]: Name of the vars should match name of the package

Comment (oivindoh) [reply template.env:10]: You want to drop the PYPI_ prefix?

Comment (ostefano) [reply template.env:10]: Nope, sorry I meant 'MISP_LIB_STIX2' for example.
Is it called like that? If so then ignore my previous comment

File: server/Dockerfile
Comment (ostefano) [reply server/Dockerfile]: Actually we should give priority to our versions IFF they are defined to non-null strings.

File: template.env
Comment (oivindoh) [reply template.env:10]: Yeah, or it's named misp-lib-stix2, but dashes are not valid in posix env var names

File: server/Dockerfile
Comment (oivindoh) [reply server/Dockerfile]: OK I've [implemented this logic](https://github.com/ostefano/docker-misp/pull/27/commits/a0e640ef881e66cd6b44de2f3ca130ba1def79c8) now.

For each module we want to add,

- If it's missing in requirements.txt, just add it, optionally with version specified
- If it's present in requirements.txt and we have a version specified, replace it with the specified version
- If it's present in requirements.txt and we do not have a version specified, leave the requirements.txt entry alone

File: docker-bake.hcl
Comment (oivindoh) [reply docker-bake.hcl]: [fixed this now](https://github.com/ostefano/docker-misp/pull/27/commits/a0e640ef881e66cd6b44de2f3ca130ba1def79c8)

File: docker-bake.hcl
Comment (ostefano) [inline docker-bake.hcl]: trailing comma here missing (better to have trailing commas so diffs are easier)

File: server/Dockerfile
Comment (ostefano) [reply server/Dockerfile]: Good logic. Can you add a comment with the explanation?

File: docker-bake.hcl
Comment (oivindoh) [reply docker-bake.hcl]: added

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: If you are into fixing existing typos, here another one "new new" :)

**Comment by ostefano** [comment]
@oivindoh thanks for your work! Left a few more comments, but I think we converged.
After your next iteration I will test things locally and merge.

**Comment by oivindoh** [comment]
Latest commit is to deal with the case of missing MISP_TAG or MODULES_TAG  after merging https://github.com/ostefano/docker-misp/pull/23

**Comment by oivindoh** [comment]
Last commit properly deals with file writability check for .git/ORIG_HEAD (just ensure it exists and is writable) and cleans up the image by removing the git application as well as /var/www/MISP/.git/ (leaving HEAD because this file is still read directly by MISP even with mentioned PR)

This slims the image down to 255MB for the arm64 version (built from `MISP_COMMIT=89fc50118fb0a7a64cef57baa3668591d05891dd`)

**Comment by oivindoh** [comment]
rebased onto master.

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: Are lines 130 and 132 still needed after the merge?

File: server/Dockerfile
Comment (ostefano) [inline server/Dockerfile]: What do we do here? Are we preparing for the copy?

File: server/files/entrypoint_nginx.sh
Comment (ostefano) [inline server/files/entrypoint_nginx.sh]: Is this needed after merging the PR in MISP?

Also, aren't we doing already in the Dockerfile? Couldn't we chmod there?

File: server/Dockerfile
Comment (oivindoh) [reply server/Dockerfile]: Yes, there is still a direct read of the HEAD file that causes an error in the log unless HEAD is present, while ORIG_HEAD is still expected to be present and writeable. I removed the part of my now merged upstream PR that excluded this file to increase the chances of getting merged (as the removal felt like it unnecessarily cluttered their code)

File: server/files/entrypoint_nginx.sh
Comment (oivindoh) [reply server/files/entrypoint_nginx.sh]: I tried Dockerfile chmod but the entrypoint_nginx.sh chmods prior to this make this file unwriteable, so I just added this single chmod rather than mess with the prior entrypoint commands.

Comment (ostefano) [reply server/files/entrypoint_nginx.sh]: Proposal: should we change this line and the line in the Docker file to `touch && chmod` and change this comment to "do the action again because...etc". I would like to somehow make it explicit that these two lines do in theory the very same thing, and we need it in the both places because reasons.

File: server/Dockerfile
Comment (oivindoh) [reply server/Dockerfile]: Sorry, this is left over from attempts to copy both these in one layer, and should be redundant. Removed it.

File: server/files/entrypoint_nginx.sh
Comment (oivindoh) [reply server/files/entrypoint_nginx.sh]: Actually, this is just a sign I'm operating on too much caffeine lately. The chmod is of course overridden in the COPY step in the target container, not all the find steps in the entrypoint that don't actually target the .git directory in any way. I'll move this touch/chmod bit to the same step as dist copy.

**Comment by ostefano** [comment]
Looks good. I will check whether we can wait for a new release of MISP before merging.

**Comment by oivindoh** [comment]
Rebased onto current master

**Review by ostefano — APPROVED** [review]

**Pull request merged**
